File: week08/18/prob2.cpp
# include <iostream>
# include <vector>

using namespace std;

class MyCalendarTwo {
public:
    vector<pair<int, int>> schedule;
    vector<pair<int, int>> overlaps;

    MyCalendarTwo() {

    }
    
    bool book(int start, int end) {
        
        for (auto& [o_s, o_e] : overlaps){
            if (start < o_e && end > o_s){
                return false;
            }
        }

        for (auto& [s_s, s_e] : schedule){
            if (start < s_e && end > s_s){
                overlaps.push_back({max(start, s_s), min(end, s_e)});
            }
        }

        schedule.push_back({start, end});
        return true;
    }
};

File: week08/26/prob2.py
class MyCalendarTwo:

    def __init__(self):
        self.events = []


    def book(self, start: int, end: int) -> bool:
        overlap_area = []
        for i, e in enumerate(self.events):
            if max(start, e[0]) < min(end, e[1]):
                overlap_area.append((max(start, e[0]), min(end, e[1])))
        for i in range(len(overlap_area)):
            s1, e1 = overlap_area[i]
            for j in range(i+1, len(overlap_area)):
                s2, e2 = overlap_area[j]
                if max(s1, s2) < min(e1, e2):
                    return False
        self.events.append([start, end])
        return True
                
            

                

            
        self.events.append([start, end])
        return True
            
        


# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)